src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketDeflater.cs

```diff
@@ -0,0 +1,229 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using static System.IO.Compression.ZLibNative;
+
+namespace System.Net.WebSockets.Compression
+{
+    /// <summary>
+    /// Provides a wrapper around the ZLib compression API.
+    /// </summary>
+    internal sealed class WebSocketDeflater : IDisposable
+    {
+        private ZLibStreamHandle? _stream;
+        private readonly int _windowBits;
+        private readonly bool _persisted;
+
+        private byte[]? _buffer;
```

Comment:
In the name of memory usage, these buffers should be as temporary as possible and should be returned to the array pool when they are consumed. A new one should be rented per operation (where possible).

Reply:
Agree 100%. This buffer is released by ManagedWebSocket when it's done writing the deflated payload into it's own send buffer. See https://github.com/zlatanov/runtime/blob/17b1e0292b4217602a0a57abb1c2437efd7220e7/src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs#L535-L588. 

This can be made more obvious if I returned IMemoryOwner<byte> and not holding to this _buffer, but that would cause an allocation for the IMemoryOwner<byte> instance.